src/server/coprocessor/executor/index_scan.rs

```diff
@@ -0,0 +1,318 @@
+            pk_col: pk_col,
+        }
+    }
+
+    pub fn get_row_from_range(&mut self) -> Result<Option<Row>> {
+        let range = &self.key_ranges[self.cursor];
+        if range.get_start() > range.get_end() {
+            return Ok(None);
+        }
+        let kv = try!(self.scanner.next_row(range));
+        let (key, value) = match kv {
+            Some((key, value)) => (key, value),
+            None => return Ok(None),
+        };
+
+        let seek_key = if self.desc {
+            key.clone()
+        } else {
+            prefix_next(&key)
+        };
+        self.scanner.set_seek_key(Some(seek_key));
+
+        let (mut values, handle) = {
+            box_try!(table::cut_idx_key(key, &self.col_ids))
+        };
+
+        let handle = if handle.is_none() {
+            box_try!(value.as_slice().read_i64::<BigEndian>())
+        } else {
+            handle.unwrap()
+        };
+
+        if let Some(ref pk_col) = self.pk_col {
+            let handle_datum = if mysql::has_unsigned_flag(pk_col.get_flag() as u64) {
+                // PK column is unsigned
+                datum::Datum::U64(handle as u64)
+            } else {
+                datum::Datum::I64(handle)
+            };
+            let mut bytes = datum::encode_value(&[handle_datum]).unwrap();
+            values.append(pk_col.get_column_id(), &mut bytes);
+        }
+        Ok(Some(Row::new(handle, values)))
+    }
+}
+
+impl<'a> Executor for IndexScanExec<'a> {
+    fn next(&mut self) -> Result<Option<Row>> {
+        while self.cursor < self.key_ranges.len() {
+            let data = box_try!(self.get_row_from_range());
+            if data.is_none() {
+                self.scanner.set_seek_key(None);
+                self.cursor += 1;
+                continue;
+            }
+            return Ok(data);
+        }
+        Ok(None)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use super::super::scanner::test::{Data, TestStore, new_col_info};
+    use util::codec::mysql::types;
+    use util::codec::datum::{self, Datum};
+    use util::codec::number::NumberEncoder;
+    use util::collections::HashMap;
+    use std::i64;
+    use tipb::schema::ColumnInfo;
+    use storage::Statistics;
+    use protobuf::RepeatedField;
+
+    const TABLE_ID: i64 = 1;
+    const INDEX_ID: i64 = 1;
+    const KEY_NUMBER: usize = 10;
+
+    #[inline]
+    pub fn get_idx_range(table_id: i64, idx_id: i64, start: i64, end: i64) -> KeyRange {
+        let mut start_buf = Vec::with_capacity(8);
+        start_buf.encode_i64(start).unwrap();
+        let mut end_buf = Vec::with_capacity(8);
+        end_buf.encode_i64(end).unwrap();
+        let mut key_range = KeyRange::new();
+        key_range.set_start(table::encode_index_seek_key(table_id, idx_id, &start_buf));
+        key_range.set_end(table::encode_index_seek_key(table_id, idx_id, &end_buf));
+        key_range
+    }
+
+    pub fn prepare_index_data(key_number: usize, table_id: i64, index_id: i64) -> Data {
```

Comment:
This is too complex for test code.
Do we really need to store `encode_data`?

Since the data is generated by our code, We can simply compute the value based on the handle of a row, we just need the handle to verify the result.